### PR TITLE
[Influxdb] Allow users to add theirs owns tags

### PR DESCRIPTION
**Pull request by max3163**
This PR allow users to add ( from the parameters list ) theirs own tags to influxdb.
By example you can add the tag environnement with rct, qal, prd value ( or acceptance, live etc... ). In this way you can have multiple test in the same time on differents environnements without confuse datas in influxdb.

image

### CODE REVIEW

File: src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java [hunk 1]
```diff
@@ -71,6 +72,10 @@
 
     /** A panel allowing the user to set arguments for this test. */
     private ArgumentsPanel argsPanel;
+    
+    /** The current className of the Backend listenenr **/
+    private String className;
```
Comment (pmouawad) [inline src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java]: This part does not seem to be related to this PR , it's probably here to avoid losing configuration when changing class. Would it be possible to make another PR ?

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java [hunk 2]
```diff
@@ -105,7 +117,7 @@
     private String testTags;
     // Name of the application tested
     private String application = "";
-
+    private String TAG_USER = "";
```
Comment (pmouawad) [inline src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Should be lower case or static final

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java [hunk 3]
```diff
@@ -318,6 +335,13 @@ public void setupTest(BackendListenerContext context) throws Exception {
                 }
             }
         }
+        // Check if more fields are filled ( corresponding to user tag )
+        context.getParameterNamesIterator().forEachRemaining(name->{
+            if ( !defaultArg.containsKey(name)) {
+                TAG_USER += "," + AbstractInfluxdbMetricsSender.tagToStringValue(name.trim()) + "=" + AbstractInfluxdbMetricsSender.tagToStringValue(context.getParameter(name).trim());
```
Comment (pmouawad) [inline src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Add some logging here to debug in case of issue

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java [hunk 4]
```diff
@@ -92,6 +92,18 @@
     private static final int MAX_POOL_SIZE = 1;
     private static final String SEPARATOR = ";"; //$NON-NLS-1$
     private static final Object LOCK = new Object();
+    private static Map<String, String> defaultArg = new LinkedHashMap<>();
```
Comment (pmouawad) [inline src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Should be UPPER CASE

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java (hunk 3 above)
Comment (pmouawad) [inline src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Thanks for this PR. It would require more documentation as I find the usage a bit edgy. AFAIU you 'll enable user to add as many rows (key=value) which will be tags for Influx event.
Can't we have some meaningful prefix or something to avoid a conflict between a user tag and a real property of the InfluxdbBackendListenerClient ?

**Review by pmouawad — CHANGES_REQUESTED** [review]
Thanks for you PR Maxime.
I have added few code comments. I think it is necessary to take them into account before commiting to svn.

Thanks

File: src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java (hunk 1 above)
Comment (max3163) [reply src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java]: Not only, for me these changes are an integral part of the PR. Without it, you can swap class without lose or merge configurations.

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java (hunk 4 above)
Comment (max3163) [reply src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Fix

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java (hunk 2 above)
Comment (max3163) [reply src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Fix

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java (hunk 3 above)
Comment (max3163) [reply src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: Done

Comment (max3163) [reply src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: I don't think the user will add a lot of tag and it's more clear for me in this way and it's easier to parse data into influxdb line protocol. 
If you want, we can add a '_' at the start of each user tag but for me it's not mandatory as real property of the InfluxdbBackendListenerClient are default setted

File: src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java (hunk 1 above)
Comment (pmouawad) [reply src/components/org/apache/jmeter/visualizers/backend/BackendListenerGui.java]: ok

File: src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java (hunk 3 above)
Comment (pmouawad) [reply src/components/org/apache/jmeter/visualizers/backend/influxdb/InfluxdbBackendListenerClient.java]: What if user names his tag the same as an existing attribute ? he can't . If we add a prefix "TAG_" then I feel it's more clear and avoid conflicts. 
Also I think this needs more documentation. Feature without documentation is not usable. We currently lack documentation on InfluxDBBackendListenerClient.
I think it's a feature that is very interesting and we need users to know it exists and what it does.

**Review by pmouawad — CHANGES_REQUESTED** [review]
Thanks for taking into account my remarks.
There are 2 remaining points:
- Documentation
- Prefix for custom tags if you agree with my arguments

Thanks

**Comment by pmouawad** [comment]
@max3163 , thanks for updating your PR.
It is ok for me now.

Regards

**Comment by pmouawad** [comment]
Resolved by commit:
http://svn.apache.org/viewvc?rev=1815893&view=rev

@max3163 , can you close this one please ?
Thank you

**Comment by hemaraoshop** [comment]
Hi,

We are trying to send some data in influxdb using the BackEnd listener's (org.apache.jmeter.visualizers.backend.influxdb.InfluxdbBackendListenerClient) new TAG_ feature in Jmeter 4.0 but we are not seeing any data going to the InfluxDb. Request you to please help us out on how to send data to Influx DB using the TAG_ feature of the Backend Listener?

**Comment by max3163** [comment]
Check if you correctly add a new row on the backend listenner like this : 

image

And check in influxdb the measurement like this : 

image

For me it's work

**Comment by hemaraoshop** [comment]
We have tried exactly as shown in above screenshot, but still we are not able to send data to influxDB.
